torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1265,20 +1265,56 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
 
 void ProcessGroupNCCL::heartbeatMonitor() {
   uint64_t heartBeatCounter = 0ULL;
+  int heartBeatTimeout = (dumpOnTimeout_ && uid_ == 0)
+      ? timeoutCheckIntervalMilSec_ * 1000
+      : heartbeatTimeoutInSec_;
+  auto lastTimePollStore = std::chrono::steady_clock::now();
+  std::future<bool> asyncDebugDump;
   while (true) {
     // This won't have any lock since this lock is only used here.
     // Please be aware that mutex `monitorMutex_` should not be used
     // somewhere else to avoid the deadlock.
     std::unique_lock<std::mutex> lock(monitorMutex_);
     if (monitorWakeUpCV_.wait_for(
-            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+            lock, std::chrono::seconds(heartBeatTimeout), [&] {
               return terminateHeartbeatMonitorThread_.load();
             })) {
       // For the normal complete or user interception, monitorWakeUpCV_
       // will get notified, we early return and exit heartbeatMonitor.
       return;
     }
 
+    // Assuming that we always init a process group containing all ranks,
```

Comment:
"We put extra functionality in the thread for the default PG (uid_=0) any time we need something to run once per process and avoid doing it from too many separate threads.  For example, we check a global flag on the TCPStore to see if anyone else observed timeout and signaled peers to dump state, and we avoid hammering the TCPStore from all threads"  might be clearer